pdns/nod.cc

```diff
@@ -0,0 +1,211 @@
+              if ((last_write_time(i->path()) < newest_time) ||
+                  newest_file.empty()) {
+                newest_time = last_write_time(i->path());
+                newest_file = i->path();
+              }
+            }
+          }
+        }
+        if (exists(newest_file)) {
+          std::string filename = newest_file.string();
+          std::ifstream infile;
+          try {
+            infile.open(filename, std::ios::in | std::ios::binary);
+            g_log << Logger::Warning << "Found SBF file " << filename << endl;
+            // read the file into the sbf
+            d_sbf.restore(infile);
+            infile.close();
+            // now dump it out again with new thread id & process id
+            snapshotCurrent();
+            // Remove the old file we just read to stop proliferation
+            remove(newest_file);
+          }
+          catch (const std::runtime_error& e) {
+            g_log<<Logger::Warning<<"NODDB init: Cannot parse file: " << filename << endl;
+          }
+        }
+      }
+    }
+    catch (const filesystem_error& e) {
+      g_log<<Logger::Warning<<"NODDB init failed:: " << e.what() << endl;
+      return false;
+    }
+  }
+  d_init = true;
+  return true;
+}
+
+void NODDB::housekeepingThread()
+{
+  for (;;) {
+    sleep(d_snapshot_interval);
+    {
+      snapshotCurrent();
+    }
+  }
+}
+
+void NODDB::setCacheDir(const std::string& cachedir)
+{
+  if (!d_init) {
+    path p(cachedir);
+    if (!exists(p))
+      throw PDNSException("NODDB setCacheDir specified non-existent directory: " + cachedir);
+    else if (!is_directory(p))
+      throw PDNSException("NODDB setCacheDir specified a file not a directory: " + cachedir);
+    d_cachedir = cachedir;
+  }
+}
+
+bool NODDB::isNewDomain(const std::string& domain)
+{
+  DNSName dname(domain);
+  return isNewDomain(dname);
+}
+
+bool NODDB::isNewDomain(const DNSName& dname)
+{
+  // The only time this should block is when snapshotting from the
+  // housekeeping thread
+  std::lock_guard<std::mutex> lock(NODDB::d_sbf_mutex);
+  // the result is always the inverse of what is returned by the SBF
+  return !d_sbf.testAndAdd(dname.toDNSStringLC());
+}
+
+bool NODDB::isNewDomainWithParent(const std::string& domain, std::string& observed)
+{
+  DNSName dname(domain);
+  return isNewDomainWithParent(dname, observed);
+}
+
+bool NODDB::isNewDomainWithParent(const DNSName& dname, std::string& observed)
+{
+  bool ret = isNewDomain(dname);
+  if (ret == true) {
+    DNSName mdname = dname;
+    while (mdname.chopOff()) {
+      if (!isNewDomain(mdname)) {
+        observed = mdname.toString();
+        break;
+      }
+    }
+  }
+  return ret;
+}
+
+void NODDB::addDomain(const DNSName& dname)
+{
+  // The only time this should block is when snapshotting from the
+  // housekeeping thread
+  std::lock_guard<std::mutex> lock(NODDB::d_sbf_mutex);
+  std::string native_domain = dname.toDNSStringLC();
```

Comment:
the call to `DNSName::toDNSStringLC()` could be done outside of the critical section.

Reply:
yes good point